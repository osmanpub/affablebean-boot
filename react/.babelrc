 
{
  "presets": ["@babel/preset-react"],
  "env": {
    "test": {
      "plugins": [
        ["istanbul", {
          "exclude": [
            "**/*.spec.js"
          ]
        }]
      ]
    }
  },
  "plugins": ["transform-class-properties", "istanbul"]
}
