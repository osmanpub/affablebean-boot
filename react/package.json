{
  "name": "affablebean",
  "version": "0.1.0",
  "private": true,
  "eslintConfig": {
    "extends": "react-app",
    "plugins": [
      "react-hooks"
    ],
    "rules": {
      "react-hooks/rules-of-hooks": "error",
      "react-hooks/exhaustive-deps": "warn"
    }
  },
  "scripts": {
    "build": "react-scripts build",
    "coveralls": "cat coverage/lcov.info | coveralls",
    "cypress:open": "cypress open",
    "cypress:run": "cypress run",
    "cypress:verify": "cypress verify",
    "eject": "react-scripts eject",
    "report:coverage": "nyc report --reporter=html",
    "report:coverage:text": "nyc report --reporter=text",
    "start": "node_modules/react-scripts/bin/react-scripts.js start &",
    "start:java": "cd .. && sh server && cd react && node_modules/react-scripts/bin/react-scripts.js start &",
    "test": "react-scripts test"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "dependencies": {
    "@reduxjs/toolkit": "^1.2.5",
    "axios": "^0.19.2",
    "node-rest-client": "^3.1.0",
    "react": "^16.12.0",
    "react-alert": "^6.0.0",
    "react-alert-template-basic": "^1.0.0",
    "react-dom": "^16.12.0",
    "react-hook-form": "^4.9.7",
    "react-redux": "^7.2.0",
    "react-router-dom": "^5.1.2",
    "react-scripts": "3.4.0",
    "redux": "^4.0.5",
    "styled-components": "^5.0.1"
  },
  "devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/preset-react": "^7.8.3",
    "@cypress/code-coverage": "^1.12.2",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "^9.4.0",
    "@testing-library/user-event": "^7.2.1",
    "@types/axios": "^0.14.0",
    "@types/cypress": "^1.1.3",
    "@types/jest": "^25.1.3",
    "@types/react": "^16.9.21",
    "@types/react-alert": "^5.2.0",
    "@types/react-bootstrap": "^0.32.20",
    "@types/react-dom": "^16.9.5",
    "@types/react-redux": "^7.1.7",
    "@types/react-router-dom": "^5.1.3",
    "@types/react-test-renderer": "^16.9.2",
    "@types/redux": "^3.6.0",
    "@types/redux-mock-store": "^1.0.2",
    "@types/styled-components": "^5.0.0",
    "babel-plugin-istanbul": "^6.0.0",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "coveralls": "^3.0.9",
    "cross-env": "^7.0.0",
    "cypress": "^4.0.2",
    "istanbul-lib-coverage": "^3.0.0",
    "nyc": "^15.0.0",
    "react-test-renderer": "^16.12.0",
    "redux-mock-store": "^1.5.4",
    "typescript": "^3.8.2"
  },
  "nyc": {
    "sourceMap": false,
    "instrument": false
  }
}
