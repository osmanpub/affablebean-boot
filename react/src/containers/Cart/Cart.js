import React, { Component } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import CartItem from "../../components/CartItem";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import { ActionBar, CartTable, ShoppingCart, Subtotal } from "./Cart.styles";
import { clearCart } from "../../actions";

export class Cart extends Component {
  componentDidUpdate(prevProps) {
    const { cart, dispatch, match } = this.props;
    const { params } = match;

    if (cart.numberOfItems > 0 && params.clear === "true") {
      dispatch(clearCart());
    }
  }

  render() {
    const { cart, dispatch, match } = this.props;
    const { numberOfItems } = cart;

    const items = cart.items.map((item, index) => (
      <CartItem
        key={index}
        cart={cart}
        dispatch={dispatch}
        index={index}
        item={item}
      />
    ));

    const clearCart =
      numberOfItems === 0 ? (
        ""
      ) : (
        <Link className="bubble hMargin" to={"/viewCart/true"}>
          clear cart
        </Link>
      );

    return (
      <div>
        <Header cart={cart} url={match.url} />
        <div className="singleColumn">
          {numberOfItems === 0
            ? "Your shopping cart is empty"
            : "Your shopping cart contains " + numberOfItems + " items"}
          <ActionBar>
            {clearCart}
            <Link className="bubble hMargin" to={"/"}>
              continue shopping
            </Link>
          </ActionBar>
          <ShoppingCart {...cart}>
            <Subtotal>
              subtotal &euro;
              {cart.subtotal.toFixed(2)}
            </Subtotal>
            <CartTable>
              <tbody>
                <tr className="header">
                  <th>product</th>
                  <th>name</th>
                  <th>price</th>
                  <th>quantity</th>
                </tr>
                {items}
              </tbody>
            </CartTable>
          </ShoppingCart>
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => {
  const { cart } = state;

  return {
    cart
  };
};

export default connect(mapStateToProps)(Cart);

Cart.propTypes = {
  cart: PropTypes.object,
  dispatch: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired,
  url: PropTypes.string
};
