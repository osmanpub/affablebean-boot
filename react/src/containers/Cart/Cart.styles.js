import styled from "styled-components";

export const EmptyCart = styled.div`
  display: ${cart => (cart.numberOfItems === 0 ? "block" : "none")};
`;
