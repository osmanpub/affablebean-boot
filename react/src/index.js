import React from "react";
import { render } from "react-dom";
import { configureStore } from "redux-starter-kit";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route } from "react-router-dom";
import reducer from "./reducers";
import App from "./containers/App";
import CategoryProducts from "./containers/CategoryProducts";
import * as serviceWorker from "./serviceWorker";

const store = configureStore({
  reducer
});

render(
  // <React.StrictMode>
  <Provider store={store}>
    <Router>
      <Route path="/" exact component={App} />
      <Route path="/category/:id" component={CategoryProducts} />
    </Router>
  </Provider>,
  // </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
