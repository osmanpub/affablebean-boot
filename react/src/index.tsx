import { configureStore } from "@reduxjs/toolkit";
import React, { lazy, Suspense } from "react";
import { render } from "react-dom";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import ErrorBoundary from "./components/ErrorBoundary";
import rootReducer from "./redux";
import * as serviceWorker from "./serviceWorker";
const Cart = lazy(() => import("./containers/Cart"));
const CategoryProducts = lazy(() => import("./containers/CategoryProducts"));
const Checkout = lazy(() => import("./containers/Checkout"));
const Contact = lazy(() => import("./containers/Contact"));
const Home = lazy(() => import("./containers/Home"));
const Privacy = lazy(() => import("./containers/Privacy"));

const store = configureStore({
  reducer: rootReducer
});

render(
  <React.StrictMode>
    <ErrorBoundary>
      <Provider store={store}>
        <Router>
          <Suspense fallback={<div>Loading...</div>}>
            <Switch>
              <Route path="/" exact component={Home} />
              <Route path="/category/:id" component={CategoryProducts} />
              <Route path="/checkout" component={Checkout} />
              <Route path="/contact" component={Contact} />
              <Route path="/privacy" component={Privacy} />
              <Route path="/viewCart/:clear" component={Cart} />
            </Switch>
          </Suspense>
        </Router>
      </Provider>
    </ErrorBoundary>
  </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
