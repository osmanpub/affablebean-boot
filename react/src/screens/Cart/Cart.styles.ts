import styled from "styled-components";
import { Cart } from "../../interfaces/cart";

export const ActionBar = styled.div`
  margin: 30px;
  width: 750px;
  text-align: center;
`;

export const CartTable = styled.table`
  width: 750px;
`;

export const CartTableTd = styled.td`
  width: 145px;
  height: 90px;
`;

export const ShoppingCart = styled.div`
  display: ${(cart: Cart) => (cart.numberOfItems > 0 ? "block" : "none")};
`;

export const Subtotal = styled.h4`
  margin: 40px 0 20px 430px;
`;
