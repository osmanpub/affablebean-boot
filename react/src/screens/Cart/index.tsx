import React, { useEffect } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import CartItem from "../../components/CartItem";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import { Cart as CartState } from "../../interfaces/cart";
import { Match } from "../../interfaces/router";
import { RootState } from "../../redux";
import { clearCart } from "../../redux/cart";
import { ActionBar, CartTable, ShoppingCart, Subtotal } from "./Cart.styles";

type Props = {
  cart: CartState;
  clearCart: Function;
  match: Match;
};

function Cart(props: Props) {
  const { cart, clearCart, match } = props;
  const { numberOfItems } = cart;

  useEffect(() => {
    const { params } = match;

    if (cart.numberOfItems > 0 && params.clear === "true") {
      clearCart();
    }
  }, [cart.numberOfItems, clearCart, match]);

  const items = cart.items.map((item, index) => (
    <CartItem key={index} index={index} item={item} />
  ));

  const clear =
    numberOfItems === 0 ? (
      ""
    ) : (
      <Link
        className="bubble hMargin"
        to={"/viewCart/true"}
        data-cy="clear-cart"
      >
        clear cart
      </Link>
    );

  return (
    <div>
      <Header url={match.url} />
      <div className="singleColumn" data-cy="cart-info">
        {numberOfItems === 0
          ? "Your shopping cart is empty"
          : "Your shopping cart contains " + numberOfItems + " items"}
        <ActionBar>
          {clear}
          <Link className="bubble hMargin" to={"/"} data-cy="continue-shopping">
            continue shopping
          </Link>
        </ActionBar>
        <ShoppingCart {...cart}>
          <Subtotal>
            subtotal &euro;
            {cart.subtotal.toFixed(2)}
          </Subtotal>
          <CartTable>
            <tbody>
              <tr className="header">
                <th>product</th>
                <th>name</th>
                <th>price</th>
                <th>quantity</th>
              </tr>
              {items}
            </tbody>
          </CartTable>
        </ShoppingCart>
      </div>
      <Footer />
    </div>
  );
}

const mapStateToProps = (state: RootState) => ({
  cart: state.cart
});

const mapDispatchToProps = {
  clearCart
};

export default connect(mapStateToProps, mapDispatchToProps)(Cart);
