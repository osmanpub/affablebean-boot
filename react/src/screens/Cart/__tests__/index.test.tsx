import { act, fireEvent, render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Cart from "..";

const mockStore = configureStore([thunk]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [
      {
        product: {
          id: 1,
          description: "semi skimmed (1L)",
          name: "milk",
          price: 1.7
        },
        quantity: 1,
        total: 1.7
      }
    ],
    numberOfItems: 1,
    subtotal: 1.7
  }
};

const store = mockStore(initialState);

const cart = (
  <Router>
    <Provider store={store}>
      <Cart
        match={{
          path: "/viewCart/:clear",
          url: "/viewCart/false",
          isExact: true,
          params: { clear: "false" }
        }}
      />
    </Provider>
  </Router>
);

describe("<Cart />", () => {
  it("renders correctly", () => {
    const component = renderer.create(cart).toJSON();
    // @ts-ignore
    expect(component).toMatchSnapshot();
  });

  it("show smaller navigatioal logo, on left", () => {
    const { getByAltText } = render(cart);
    const logo = getByAltText(/Affable Bean logo/i);
    // @ts-ignore
    expect(logo).toBeInTheDocument();
  });

  it("show larger logo, on right", () => {
    const { getByAltText } = render(cart);
    const logo = getByAltText(/the affable bean/i);
    // @ts-ignore
    expect(logo).toBeInTheDocument();
  });

  it("show proceed to checkout link", () => {
    const { getByText } = render(cart);
    const link = getByText(/proceed to checkout/i);
    // @ts-ignore
    expect(link).toBeVisible();
  });

  it("cart header total has 1 item", () => {
    const { getAllByText } = render(cart);
    const items = getAllByText(/1 items/i);
    // @ts-ignore
    expect(items[0]).toBeVisible();
  });

  it("show privacy", () => {
    const { getByTestId } = render(cart);
    const privacy = getByTestId(/footer-privacy/i);
    // @ts-ignore
    expect(privacy).toBeInTheDocument();
  });

  it("show contact", () => {
    const { getByTestId } = render(cart);
    const contact = getByTestId(/footer-contact/i);
    // @ts-ignore
    expect(contact).toBeInTheDocument();
  });
});

describe("<CartItem />", () => {
  it("update quantity to 2", () => {
    const { getByDisplayValue, getByText } = render(cart);
    const input = getByDisplayValue("1");
    const updateButton = getByText(/update/i);

    act(() => {
      fireEvent.change(input, { target: { value: "2" } });
      fireEvent.click(updateButton);
    });

    // @ts-ignore
    expect(input.value).toBe("2");
  });
});
