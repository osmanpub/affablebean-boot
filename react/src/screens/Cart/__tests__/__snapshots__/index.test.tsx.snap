// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders correctly 1`] = `
<div>
  <div
    className="sc-AxheI igDRXH"
  >
    <div
      className="sc-Axmtr iitsqH"
    >
      <div
        className="sc-AxhCb kdYhXy"
      >
        <div
          className="sc-AxhUy bLXZsm headerWidget"
        >
          <a
            className="bubble"
            data-cy="proceed-checkout"
            href="/checkout"
            onClick={[Function]}
          >
            proceed to checkout ➟
          </a>
        </div>
        <div
          className="sc-AxgMl kJNXLs headerWidget"
        >
          <span
            className="horizontalMargin"
            data-cy="cart-hdr-total"
            data-testid="cart-hdr-total"
          >
            <img
              alt="shopping cart icon"
              src="/static/img/cart.gif"
            />
             
            1 items
          </span>
          
        </div>
      </div>
      <div
        className="sc-AxmLO bFONzf"
      >
        <a
          data-cy="home-logo"
          href="/"
          onClick={[Function]}
        >
          <img
            alt="Affable Bean logo"
            src="/static/img/logo.png"
          />
        </a>
      </div>
      <img
        alt="the affable bean"
        id="logoText"
        src="/static/img/logoText.png"
      />
    </div>
  </div>
  <div
    className="singleColumn"
    data-cy="cart-info"
  >
    Your shopping cart contains 1 items
    <div
      className="sc-fzozJi vlZWm"
    >
      <a
        className="bubble hMargin"
        data-cy="clear-cart"
        href="/viewCart/true"
        onClick={[Function]}
      >
        clear cart
      </a>
      <a
        className="bubble hMargin"
        data-cy="continue-shopping"
        href="/"
        onClick={[Function]}
      >
        continue shopping
      </a>
    </div>
    <div
      className="sc-fzplWN czluRe"
    >
      <h4
        className="sc-fznyAO dAtAtF"
      >
        subtotal €
        1.70
      </h4>
      <table
        className="sc-fzoLsD gXeYhr"
      >
        <tbody>
          <tr
            className="header"
          >
            <th>
              product
            </th>
            <th>
              name
            </th>
            <th>
              price
            </th>
            <th>
              quantity
            </th>
          </tr>
          <tr
            className="white"
            data-cy="product-milk"
          >
            <td
              className="sc-AxjAm jhCEof"
            >
              <img
                alt="milk"
                src="/static/img/products/milk.png"
              />
            </td>
            <td
              className="sc-AxjAm jhCEof"
              data-testid="milk"
            >
              milk
            </td>
            <td
              className="sc-AxjAm jhCEof"
              data-testid={1.7}
            >
              €
              1.70
              <br />
              1.7
            </td>
            <td
              className="sc-AxjAm jhCEof"
            >
              <form
                onSubmit={[Function]}
              >
                <input
                  className="form-control"
                  data-cy="input-qty-milk"
                  defaultValue={1}
                  name="quantity"
                  size={2}
                  style={
                    Object {
                      "margin": "5px",
                      "textAlign": "center",
                      "width": "90%",
                    }
                  }
                  type="number"
                />
                <button
                  className="btn btn-primary btn-sm"
                  data-cy="update-qty-milk"
                  type="submit"
                >
                  update
                </button>
              </form>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="sc-AxirZ MueCz"
  >
    <br />
    <span
      className="sc-AxiKw gXNbfu"
    >
      <a
        data-cy="privacy"
        data-testid="footer-privacy"
        href="/privacy"
        onClick={[Function]}
      >
        <button
          className="btn btn-link"
        >
          Privacy
        </button>
      </a>
      <a
        data-cy="contact"
        data-testid="footer-contact"
        href="/contact"
        onClick={[Function]}
      >
        <button
          className="btn btn-link"
        >
          Contact
        </button>
      </a>
       © 2020 the affable bean company
    </span>
  </div>
</div>
`;
